// israeli-bank-scrapers-core/lib/helpers/navigation.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getCurrentUrl = getCurrentUrl;
exports.waitForNavigation = waitForNavigation;
exports.waitForNavigationAndDomLoad = waitForNavigationAndDomLoad;
exports.waitForRedirect = waitForRedirect;
exports.waitForUrl = waitForUrl;
require("core-js/modules/es.promise.js");
require("core-js/modules/es.regexp.constructor.js");
require("core-js/modules/es.regexp.exec.js");
var _waiting = require("./waiting");
async function waitForNavigation(pageOrFrame, options) {
  await pageOrFrame.waitForNavigation(options);
}
async function waitForNavigationAndDomLoad(page) {
  await waitForNavigation(page, {
    waitUntil: 'domcontentloaded'
  });
}
function getCurrentUrl(pageOrFrame, clientSide = false) {
  if (clientSide) {
    return pageOrFrame.evaluate(() => window.location.href);
  }
  return pageOrFrame.url();
}
async function waitForRedirect(pageOrFrame, timeout = 20000, clientSide = false, ignoreList = []) {
  const initial = await getCurrentUrl(pageOrFrame, clientSide);
  await (0, _waiting.waitUntil)(async () => {
    const current = await getCurrentUrl(pageOrFrame, clientSide);
    return current !== initial && !ignoreList.includes(current);
  }, `waiting for redirect from ${initial}`, timeout, 1000);
}
async function waitForUrl(pageOrFrame, url, timeout = 20000, clientSide = false) {
  await (0, _waiting.waitUntil)(async () => {
    const current = await getCurrentUrl(pageOrFrame, clientSide);
    return url instanceof RegExp ? url.test(current) : url === current;
  }, `waiting for url to be ${url}`, timeout, 1000);
}
